Add PlutusLedgerApi QuickCheck instances

t4ccer opened the pull request:
Closes #6210

Add `Arbitrary`, `CoArbitrary` and `Function` instances from `QuickCheck` for every `PlutusLedgerApi` type. This is a port of instances from https://github.com/Plutonomicon/plutarch-plutus

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

t4ccer committed:
Add V1 PlutusLedgerApi QuickCheck instances

t4ccer committed:
Add V2 PlutusLedgerApi QuickCheck instances

t4ccer committed:
Add V3 PlutusLedgerApi QuickCheck instances

t4ccer committed:
Remove obsolete `Arbitrary` instances

t4ccer committed:
Fix build with GHC8

t4ccer committed:
Add changelog fragment

t4ccer committed:
Remove main changelog entry

kozross commented:
@effectfully - would you mind taking a look at this?

effectfully commented:
It's on my list, but I can't do it this week, sorry, too many commitments. I'll do my best to review it the next week. Thanks for the reminder.

t4ccer committed:
Add UnsortedAssocMap

effectfully reviewed (commented):
Shrinking is broken, coverage is poor, invariants aren't respected. 3k lines of code.

Thank you for opening the PR! We'll process it, split into smaller PRs, fix issues etc.

kozross commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Common/QuickCheck/Utils.hs:
```diff
@@ -0,0 +1,110 @@
+
+instance KnownNat n => Arbitrary (SizedByteString n) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
```
This guarantees that you get `ByteString`s that are UTF-8 encodings of text, which is probably what you _often_ want. However, in a lot of situations here, we're generating arbitrary hashes, which most certainly do _not_ need to follow this rule.

effectfully replied:
Then we should change this rule in our code in the first place. Thanks!

kozross commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/V1/Contexts.hs:
```diff
@@ -0,0 +1,191 @@
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
+    TxInfo . getNonEmpty
+      <$> arbitrary -- inputs
+      <*> (getNonEmpty <$> arbitrary) -- outputs
```
Encoding such a relationship into an `Arbitrary` instance is surprisingly hard.

effectfully replied:
Makes me wonder if `ledger` folks do it, maybe we could somehow depend on their stuff for testing purposes.

kozross replied:
Wait, there are `Arbitrary` instances in the ledger? Genuine news to me.

kozross commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/V1/Contexts.hs:
```diff
@@ -0,0 +1,191 @@
+      <*> arbitrary -- tid
+
+  {-# INLINEABLE shrink #-}
+  shrink (TxInfo ins outs fee mint dcert wdrl validRange sigs dats tid) = do
+    NonEmpty ins' <- shrink (NonEmpty ins)
+    NonEmpty outs' <- shrink (NonEmpty outs)
```
Every day I learn _something_ new about QuickCheck.

effectfully replied:
Yeah, it's full of footguns, you can't trust anything.

kozross commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/V1/Interval.hs:
```diff
@@ -0,0 +1,229 @@
+  liftShrink shrinkInner = \case
+    NegInf -> []
+    Finite x -> Finite <$> shrinkInner x
+    PosInf -> []
```
I guess you're suggesting a 'towards negative infinity' shrinking strategy?

effectfully replied:
Towards `Finite 0`.

kozross replied:
Yeah, that would work, with the right constraints on the type variable.

kozross commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/PlutusTx.hs:
```diff
@@ -0,0 +1,258 @@
+  deriving (Eq, Ord) via PlutusTx.BuiltinByteString
+  deriving stock (Show)
+
+-- No shrinker, as it doesn't make much sense to.
+instance Arbitrary Blake2b244Hash where
```
Hashes are supposed to 'digest' information into a fixed size: about the only property they maintain is 'different hash implies different value'. This means that in practice, talking about a 'simpler' hash doesn't really make a lot of sense: if your tests pass on _some_ hash values but not others, this is _extremely_ surprising, as this suggests you're writing functionality against properties not inherent to hashes.

It doesn't hurt to shrink towards all-zeroes, but I can't see much benefit here.

effectfully replied:
> if your tests pass on some hash values but not others, this is extremely surprising

Not really, maybe you use the same hash in two different places. So if you try to shrink it in one, the property may stop failing and then seeing the non-zero hash will give you an idea that the issue may be related to that specific hash in the context of the whole `TxInfo` or whatever.

kozross replied:
We can add such a shrink easily.

Unisay commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/PlutusTx.hs:
```diff
@@ -0,0 +1,258 @@
+                        oneof, scale, shuffle, sized, variant)
+import Test.QuickCheck.Instances.ByteString ()
+
+instance Arbitrary PlutusTx.BuiltinByteString where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = PlutusTx.toBuiltin @ByteString <$> arbitrary
+  {-# INLINEABLE shrink #-}
+  shrink = fmap (PlutusTx.toBuiltin @ByteString) . shrink . PlutusTx.fromBuiltin
```
So this instance delegates to the `Arbitrary ByteString` instance. 
Assuming the latter is fine, this one should be also fine.

t4ccer committed:
Shrink fields one by one

t4ccer committed:
Shrink hashes towards all zeros

t4ccer committed:
Implement `Arbitrary BuiltinData` in terms of `Data`

t4ccer committed:
Shrink `Extended` towards `Finite 0`

t4ccer committed:
Generate arbitrary closure bounds at infinities

Such intervals are provided by cardano-node, see: <https://github.com/IntersectMBO/plutus/blob/c082e28591990a14a9f4f91700b684590ac252be/plutus-ledger-api/src/PlutusLedgerApi/V1/Interval.hs#L369-L393>

t4ccer commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/V1/Credential.hs:
```diff
@@ -0,0 +1,90 @@
+  arbitrary =
+    oneof
+      [ PubKeyCredential <$> arbitrary
+      , ScriptCredential <$> arbitrary
```
Yeah, you're right, but are we trying to make them match? We're not trying to generate "legal" contexts as it's _insanely_ tricky anywhere else either

t4ccer commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/V1/Value.hs:
```diff
@@ -0,0 +1,399 @@
+this can have zero values, and does not treat the Ada symbol or token name
+specially.
+-}
+instance Arbitrary Value where
```
Should we instead use instance that was in `plutus-ledger-api/testlib/PlutusLedgerApi/Test/V1/Value.hs`?

effectfully replied:
We should merge ours with yours as both have good and bad things about them. I assume MLabs isn't paid by IOG to do this work?

t4ccer replied:
No, current Plutarch work is part of Catalyst project

effectfully replied:
Good to know, thanks. It'll probably take us a lot of time to adopt all of this code tbh, due to other things being considered higher priority. We'll tag you on relevant PRs if we have any improvement suggestions over what you have here. Thanks again for creating this PR, we'll keep it open for now and we'll see how it goes with adopting your code.

t4ccer committed:
Generate large lovelace entries in `FeeValue`

t4ccer commented on plutus-ledger-api/testlib/PlutusLedgerApi/Test/Orphans/V1/Value.hs:
```diff
@@ -0,0 +1,399 @@
+described above: failing to do so means all guarantees of this type are off
+the table.
+-}
+newtype MintValue = MintValue Value
```
Could you point me to that please?

effectfully replied:
`PlutusLedgerApi.V3.MintValue.MintValue` and the generator is in `PlutusLedgerApi.Test.V3.MintValue`. Ours is probably broken though., we need to fix it.

t4ccer committed:
Merge branch 'master' into t4/quickcheck

t4ccer committed:
Fix post-merge duplicate Arbitrary instances

t4ccer committed:
Fix V3.MintValue generator

This is a backport of <https://github.com/plutonomicon/plutarch-plutus/commit/6c8db66ac8f5b2a62390503c09c0b2a9d1311035>
